src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs

```diff
@@ -103,14 +117,10 @@ public IEnumerable<ProjectId> GetProjectsWithDiagnostics()
                                                            .Select(kv => kv.Key)));
             }
 
+            [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/34761", AllowCaptures = false, AllowGenericEnumeration = false)]
             public IEnumerable<DocumentId> GetDocumentsWithDiagnostics(ProjectId projectId)
```

Comment:
we should investigate if we can make it so that instead of having these methods allocate collections which are passed back, instead have the caller create and pass in the collection to fill. The latter is often better because hte caller can appropriately pool the collections.

Reply:
Changed it so the caller pools and reuses the hashset locally.